Make it possible to select supported devices in the operator

ozhuraki opened the pull request:
In order to support per device deployment in the operator:
- Add customization overlays for FPGA, DSA, SGX, GPU, QAT under deployments/operator/device/
- Add --device cmdline to operator main.go which defines the controllers/webhooks to set up

mythi commented:
/cc @dmitsh FYI

mythi commented on cmd/operator/main.go:
```diff
@@ -95,40 +99,51 @@ func main() {
+	}
 
 	mgr.GetWebhookServer().Register("/pods", &webhook.Admission{
 		Handler: admission.HandlerFunc(pm.GetPodMutator()),
```
this is also part of FPGA.

ozhuraki replied:
OK, thanks, moved it under "if".

mythi commented on deployments/operator/device/dsa/kustomization.yaml:
```diff
@@ -0,0 +1,5 @@
+bases:
+ - ../../default
```
this also pulls in `../../crd/bases` which creates the CRDs for all devices.

ozhuraki replied:
What would be a way to drop the unneeded CRDs?

I tried with the approach to pick only specific CRDs (for example for FPGA), however couldn't get it to work:
```
diff --git a/deployments/operator/device/fpga/kustomization.yaml b/deployments/operator/device/fpga/kustomization.yaml
index cea26f2..60de267 100644
--- a/deployments/operator/device/fpga/kustomization.yaml
+++ b/deployments/operator/device/fpga/kustomization.yaml
@@ -1,5 +1,16 @@
+namespace: inteldeviceplugins-system
+namePrefix: inteldeviceplugins-
+
+resources:
+- deviceplugin.intel.com_fpgadeviceplugins.yaml
+- fpga.intel.com_acceleratorfunctions.yaml
+- fpga.intel.com_fpgaregions.yaml
+
 bases:
- - ../../default
+- ../../rbac
+- ../../manager
+- ../../webhook
+- ../../certmanager
 
 patchesStrategicMerge:
  - fpga.yaml
```

```
# kustomize build deployments/operator/device/fpga
Error: no matches for IdId apps_v1_Deployment|inteldeviceplugins-system|inteldeviceplugins-controller-manager; failed to find unique target for patch apps_v1_Deployment|inteldeviceplugins-controller-manager
```

mythi replied:
> What would be a way to drop the unneeded CRDs?

maybe the only way to do it is to have the CRDs created by the manager itself. This could follow the idea I demonstrate in https://github.com/intel/intel-device-plugins-for-kubernetes/issues/394#issuecomment-785338098

mythi commented on deployments/operator/device/dsa/dsa.yaml:
```diff
@@ -0,0 +1,14 @@
+kind: Deployment
+metadata:
+  name: inteldeviceplugins-controller-manager
+  namespace: inteldeviceplugins-system
```
was the namespace required for this to work?

ozhuraki replied:
I have tried to omit it, however kustomize errors without namespace:
```
Error: no matches for IdId apps_v1_Deployment|~X|inteldeviceplugins-controller-manager; failed to find unique target for patch apps_v1_Deployment|inteldeviceplugins-controller-manager
```

bart0sh commented on cmd/operator/main.go:
```diff
@@ -55,14 +59,18 @@ func init() {
 	flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
 		"Enable leader election for controller manager. "+
 			"Enabling this will ensure there is only one active controller manager.")
+	flag.StringVar(&deviceType, "device", "", "Device type to set up.")
```
does this mean that only one device can be chosen?

ozhuraki replied:
Sure, we can allow multiple devices. However as kustmizations overlays are per device type, would it make sense, what do you think?

bart0sh replied:
I don't know if this is possible, but it would be nice to support more than one device at a time. Otherwise if cluster has more than one device type users will have to use default operator with support of all devices.

ozhuraki replied:
OK, let's support also more then one device at a time.

ozhuraki replied:
Sure, let's make it so.

bart0sh replied:
I'd suggest a bit more user friendly approach:
```
package main
  
import "fmt"
import "flag"
import "strings"

type flagList []string

func (fl *flagList) String() string {
        return strings.Join(*fl, ", ")
}

func (fl *flagList) Set(value string) error {
        *fl = append(*fl, value)
        return nil
}

var devices flagList

func main() {
        flag.Var(&devices, "device", "device to enable, can be used multiple times")
        flag.Parse()
        fmt.Println(devices)
}
```

bart0sh replied:
sorry, forgot to show how it looks in the command line:
```
$ go run flags.go --device fpga --device gpu
[fpga gpu]
```

ozhuraki replied:
OK, thanks, let's do so.

mythi replied:
+1

bart0sh commented on cmd/operator/main.go:
```diff
@@ -95,40 +103,51 @@ func main() {
+	}
+
+	for _, device := range plugins {
 		if err = setupControllerAndWebhook[device](mgr, ns, withWebhook); err != nil {
```
This looks like we're calling setupControllerAndWebhook for all plugins. Should we do it only for chosen plugins?

ozhuraki commented on deployments/operator/device/qat/qat.yaml:
```diff
@@ -0,0 +1,14 @@
+      - args:
+        - --metrics-addr=127.0.0.1:8080
+        - --enable-leader-election
+        - --device=fpga
```
Thanks, good catch. Fixed both.

bart0sh commented on README.md:
````diff
@@ -198,6 +198,15 @@ NAME                     DESIRED   READY   NODE SELECTOR   AGE
+```bash
+$ kubectl apply -k deployments/operator/device/fpga
+```
+
````
Should we explain here how to deploy operator with more than one enabled plugin?

ozhuraki replied:
Thanks! Done.

bart0sh replied:
It's still unclear to me how to deploy more than one plugin. For example if I want to deploy operator with fpga and gpu support which path should I use for ```kubectl apply -k``` ? Where should I add ```--device fpga --device gpu``` command line options?

ozhuraki replied:
OK, updated. Is this better?

bart0sh replied:
yep, much better, thanks.

bart0sh commented on cmd/operator/main.go:
```diff
@@ -54,19 +55,46 @@ func init() {
+}
+
+func enabled(arr []string, val string) bool {
+	for _, s := range arr {
```
nit: if we change flagList []string => flagMap map[string]bool we can avoid iterating over the list of plugins.
this function would look something like this:
```
func (fm *flagMap) enabled(key string) bool {
    return (*fm)[key]
}
```

ozhuraki replied:
OK, good idea, thanks!

mythi replied:
@ozhuraki can you make this change still?

ozhuraki replied:
@mythi 

We tried to do it with map as @bart0sh suggested. However we later looked at it, @bart0sh changed his mind and we changed it back to the list, since the list approach looked slightly simpler. It's not principal, I can still change, shall we?

mythi replied:
@ozhuraki I had missed that change and this conversation was still open. I'm OK with this approach too but I'm not sure how it's simpler :-)

bart0sh commented on cmd/operator/main.go:
```diff
@@ -54,19 +55,46 @@ func init() {
 
+	if len(devices) == 0 {
+		devices = []string{"dsa", "fpga", "gpu", "qat", "sgx"}
+	}
```
this would not be needed as well if we use a map as we can initialise it with all supported plugins:
```
var devices = flagMap{"gpu": false, "fpga": false, "sgx": false, "dsa": false, "vpu": false}
```

bart0sh replied:
BTW, this approach would help to invalidate incorrect user input, e.g. if user uses "--device unsupported" we can throw an error.

bart0sh commented:
@ozhuraki There was some activity on combining components in kustomize last year: https://github.com/kubernetes-sigs/kustomize/issues/1251 Would this approach help us to generate combined operator resources for multiple devices?

ozhuraki commented:
@bart0sh 

> Would this approach help us to generate combined operator resources for multiple devices?

Thanks. It might help when we solve the problem of dropping the unwanted CRDs (at the moment they are combined).

I tried to pull only the required bases, however it didn't work (see the discussion above). One possibility is:

> maybe the only way to do it is to have the CRDs created by the manager itself. This could follow the idea I demonstrate in #394 (comment)

bart0sh commented on cmd/operator/main.go:
```diff
@@ -95,40 +114,48 @@ func main() {
 			os.Exit(1)
 		}
 	}
 
```
We should either validate --device parameter or just stop at this point if we haven't called setupContollerAndWebhook at the end of the loop.

bart0sh commented on cmd/operator/main.go:
```diff
@@ -95,40 +114,48 @@ func main() {
-	}).SetupWithManager(mgr); err != nil {
-		setupLog.Error(err, "unable to create controller", "controller", "FpgaRegion")
-		os.Exit(1)
+	if devices["fpga"] {
```
this if condition and above ```if devices["fpga"]``` can be merged into one. The same can be done for 2 ```if devices["sgx"]``` conditions.

ozhuraki replied:
OK, thanks! I will merge them.

bart0sh commented on cmd/operator/main.go:
```diff
@@ -95,40 +114,48 @@ func main() {
-	// TODO(mythi): add a StringVar flag to select which controllers to enable
-	for _, device := range []string{"dsa", "fpga", "gpu", "qat", "sgx"} {
+	for device, enabled := range devices {
+		if !enabled {
```
Oops, this condition will never be true I guess, which makes it useless to use type flagMap map[string]bool.

ozhuraki replied:
Yes, I noticed the same. The string array fits here better. OK, I will change it back.

ozhuraki commented:
OK, updated, please take a look:
- Added a validation of --device arguments
- Added a check for duplicates
- Switched to string array
- Merged "fpga" clauses

bart0sh reviewed (dismissed):
/lgtm

bart0sh commented:
@mythi @rojkov please review.

mythi commented on README.md:
````diff
@@ -198,6 +198,20 @@ NAME                     DESIRED   READY   NODE SELECTOR   AGE
+$ kubectl apply -k deployments/operator/device/fpga
+```
+
+Operator also supports deployments with multiple selected device types.
````
Let's add a note that the CRDs are created but no controller is registered.

Also, please rebase. This section got moved to `cmd/operator/README`

ozhuraki replied:
@mythi 

OK, thanks!

mythi reviewed (changes requested):
@ozhuraki sorry for the delay. I added a couple of comments.

marking as 'request changes' from me.
